### PR TITLE
OTel instrumentation

### PR DESCRIPTION
**Pull request by estolfo**
This PR complements [a PR in the elasticsearch-ruby repo](https://github.com/elastic/elasticsearch-ruby/pull/2179) to extend the way the transport layer is called by including the endpoint id and path params for a given call to `perform_request`. It also adds the creation of an OTel span for each request.

The span attributes captured follow the semantic conventions proposed in [this PR](https://github.com/open-telemetry/semantic-conventions/pull/23).

This PR does the following:
- Defines a `OpenTelemetry` class that encapsulates everything having to do with otel instrumentation in the transport layer
- Checks if the `OpenTelemetry` namespace is defined and creates a `@otel` object on the client
- Adds the `opentelemetry-sdk` to the Gemfile
- Requires a ENV variable `TEST_WITH_OTEL` that should be set to `true` if the `opentelemetry-sdk` gem should be required and the otel tests should be run
- Creates an OTel span for each request to Elasticsearch

ToDo:

- [x] Only set the request body for search-type requests / define what a search-type request is
- [x] Add sanitization of the request body when it's set on the span, with a ENV variable to configure it
- [x] Figure out how to remove the otel namespace so spans aren't created in all tests when `TEST_OTEL=true`
- [x] Update opentelemetry tests  to successfully run with ES available, if needed
- [x] Add GitHub action to run tests with `TEST_OTEL=true`
- [ ] Inline documentation
- [x] General documentation (See [this PR](https://github.com/open-telemetry/semantic-conventions/pull/23) for docs)
- [x] configuration setting to enable/disable

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -91,3 +91,15 @@ def self.included(context)
   config.formatter = 'documentation'
   config.color = true
 end
+
+if ENV['TEST_OTEL'] == 'true'
```
Comment (estolfo) [inline spec/spec_helper.rb:95]: also check if Ruby version >= 3.0? OTel requires version >= 3

File: spec/elastic/transport/opentelemetry_spec.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    end
+
+    let(:otel) { described_class.new }
+
+    context 'when path parameters' do
+      it 'creates a span with path parameters' do
+        client.perform_request(
+          'DELETE', '/foo,bar/_aliases/abc,xyz', nil, nil, nil, ["/{index}/_alias/{name}", "/{index}/_aliases/{name}"],
+          'indices.delete_alias'
+        )
```
Comment (estolfo) [inline spec/elastic/transport/opentelemetry_spec.rb:40]: @picandocodigo can you tell me how you start ES so that this request can succeed against a test server?

Comment (picandocodigo) [reply spec/elastic/transport/opentelemetry_spec.rb:40]: @estolfo For testing locally I use [the docker rake](https://github.com/elastic/elasticsearch-ruby/blob/main/CONTRIBUTING.md#tests) task in `elasticsearch-ruby`.

File: spec/elastic/transport/opentelemetry_spec.rb [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      end
+
+      context 'a non-search endpoint' do
+        let(:body) do
+          { query: { match: { something: "test" } } }
+        end
+
+        it 'does not capture db.statement' do
+          client.perform_request(
+            'POST', 'foo/_delete_by_query', nil, body, nil, ["/{index}/_delete_by_query"], 'delete_by_query'
```
Comment (estolfo) [inline spec/elastic/transport/opentelemetry_spec.rb:137]: @picandocodigo do you know of a request I could do in this spec that:
- has a body
- is a non-search endpoint
- doesn't require any setup (like index creation)

This tests fails because there is no `foo` index

File: spec/spec_helper.rb (hunk 1 above)
Comment (picandocodigo) [reply spec/spec_helper.rb:95]: Since 2.7 is already end of life'd, and we support only maintained Ruby versions, it wouldn't be necessary. I will update [the value on the gemspec file](https://github.com/elastic/elastic-transport-ruby/blob/main/elastic-transport.gemspec#L45) though to specify it requires `>= 3.0`.

Comment (estolfo) [reply spec/spec_helper.rb:95]: ok, that simplifies things. Thanks!

File: spec/elastic/transport/opentelemetry_spec.rb (hunk 3 above)
Comment (picandocodigo) [reply spec/elastic/transport/opentelemetry_spec.rb:137]: It looks like you can use `_all` instead of `foo`in this case. I tested this locally in a new cluster and it returns ok:
`client.perform_request('POST', '_all/_delete_by_query', {}, body, nil)`
The `perform_request` failed if I sent `nil` for parameters.

**Comment by estolfo** [comment]
@picandocodigo How do you suggest running these tests with the OTel instrumentation? The client is instrumented if the `opentemeltry-sdk` gem is required (i.e. the `OpenTelemetry` module is defined). [I check for an ENV variable](https://github.com/elastic/elastic-transport-ruby/pull/54/files#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94R95) `TEST_OTEL=true` in the spec helper to require the open telemetry gem.
I don't want to have OTel  instrumentation on by default for the tests so we don't slow them down but it would be nice to test the entire suite with instrumentation on, in case there are some edge cases with endpoint requests. So what do you think about doing a separate test run of the suite with `TEST_OTEL=true`?

**Comment by picandocodigo** [comment]
> So what do you think about doing a separate test run of the suite with `TEST_OTEL=true`?

Sounds good! We can definitely add a new set of tests on GitHub Actions with the env variable :+1:

File: spec/spec_helper.rb (hunk 1 above)
Comment (estolfo) [reply spec/spec_helper.rb]: just documenting here that we can't drop 2.7 for now, so we'll keep in the checks for ruby version

File: lib/elastic/transport/client.rb [hunk 4]
```diff
@@ -164,14 +165,37 @@ def initialize(arguments = {}, &block)
...
+                span["db.elasticsearch.path_parts.#{k}"] = v.join(',')
+              else
+                span["db.elasticsearch.path_parts.#{k}"] = v
+              end
+            end
+            if body_as_json = @otel.process_body(body, opts[:endpoint])
+              span['db.statement'] = body_as_json
+            end
+            span['db.operation'] = opts[:endpoint] if opts[:endpoint]
+            transport.perform_request(method, path, params || {}, body, headers)
```
Comment (picandocodigo) [inline lib/elastic/transport/client.rb:194]: Since `params` is being passed with the default value of `{}` in the `perform_request` signature, I think we don't need `|| {}` here and [on line 197](https://github.com/elastic/elastic-transport-ruby/pull/54/files#diff-78264cee2af175e6fd5eed7e26abeade1785739ab1d3ac97cf41f26ac128361bR197), right?.

Comment (estolfo) [reply lib/elastic/transport/client.rb:194]: With my testing, there was sometimes a case when there were no params but there was a body so the `perform_request` method would be called with the params arg explicitly set to `nil`. So once it gets to [this line](https://github.com/elastic/elastic-transport-ruby/blob/main/lib/elastic/transport/transport/base.rb#L283) you get an `NoMethodError: undefined method `delete' for nil:NilClass` error.

I think this scenario is already possible before these changes:
```ruby
$ client.perform_request('POST', '/', nil, '{"foo":"bar"}', '{"Content-Type":"application/x-ndjson"}')
NoMethodError: undefined method `delete' for nil:NilClass
from ..../elastic-transport-ruby/fork/elastic-transport-ruby/lib/elastic/transport/transport/base.rb:283:in `perform_request'
```

**Comment by estolfo** [comment]
Hi @picandocodigo the outstanding questions I had were:

1. Can you instantiate different OTel SDKs per process and if so, would there be a use case for passing the sdk into the client for use. This is a valid use case for Java but I confirmed with the OTel Ruby group that they have a single, global SDK.
2. Should we define a `schema_url` on the tracer? As of the time of this PR, you cannot. [Here](https://github.com/open-telemetry/opentelemetry-ruby/issues/969) is the issue in their repo.

**Comment by estolfo** [comment]
Update on this since my last comment:
I found out that [it's recommended](https://opentelemetry.io/docs/concepts/instrumentation/libraries/#getting-a-tracer) to allow a OpenTelemetry TracerProvider be specified. So I added an option when creating a client `opentelemetry_tracer_provider` in [this commit](https://github.com/elastic/elastic-transport-ruby/pull/54/commits/9a97a9c5b40c2d60f881132d77e817852e7157ea).
